File: src/lexer.rs
use crate::token::{LexerError, LexerOutput, Token};

pub struct Lexer;

impl Lexer {
    /// Determina si un lexema es una palabra reservada y devuelve su tipo de token.
    pub fn check_keyword(lexeme: &str) -> Option<&'static str> {
        match lexeme {
            // Estructuras de control
            "if" => Some("KW_IF"),
            "else" => Some("KW_ELSE"),
            "while" => Some("KW_WHILE"),
            "for" => Some("KW_FOR"),
            "do" => Some("KW_DO"),
            "return" => Some("KW_RETURN"),
            "break" => Some("KW_BREAK"),
            "continue" => Some("KW_CONTINUE"),
            "switch" => Some("KW_SWITCH"),
            "case" => Some("KW_CASE"),
            "default" => Some("KW_DEFAULT"),

            // Declaraciones y tipos de datos
            "let" => Some("KW_LET"),
            "var" => Some("KW_VAR"),
            "const" => Some("KW_CONST"),
            "fn" => Some("KW_FN"),
            "function" => Some("KW_FUNCTION"),
            "int" => Some("KW_INT"),
            "float" => Some("KW_FLOAT"),
            "double" => Some("KW_DOUBLE"),
            "char" => Some("KW_CHAR"),
            "string" => Some("KW_STRING"),
            "bool" => Some("KW_BOOL"),
            "void" => Some("KW_VOID"),
            "struct" => Some("KW_STRUCT"),
            "class" => Some("KW_CLASS"),

            // Literales booleanos / nulos
            "true" => Some("KW_TRUE"),
            "false" => Some("KW_FALSE"),
            "null" => Some("KW_NULL"),

            // Funciones / identificadores clave comunes
            "print" => Some("KW_PRINT"),
            "println" => Some("KW_PRINTLN"),
            "main" => Some("KW_MAIN"),

            _ => None,
        }
    }

    pub fn analyze(input: &str) -> LexerOutput {
        let mut tokens = Vec::new();
        let mut errors = Vec::new();

        let chars: Vec<char> = input.chars().collect();
        let len = chars.len();

        let mut i = 0;
        let mut line = 1;
        let mut col = 1;

        while i < len {
            let c = chars[i];

            // 1. Manejo de saltos de línea y espacios en blanco
            if c == '\n' {
                line += 1;
                col = 1;
                i += 1;
                continue;
            } else if c.is_whitespace() {
                col += 1;
                i += 1;
                continue;
            }

            // 2. Reconocimiento de Identificadores (ID) y Palabras Reservadas (KW_*)
            if c.is_alphabetic() || c == '_' {
                let start_col = col;
                let mut ident = String::new();

                while i < len && (chars[i].is_alphanumeric() || chars[i] == '_') {
                    ident.push(chars[i]);
                    i += 1;
                    col += 1;
                }

                // Discriminación: Palabra Reservada vs Identificador
                let token_type = match Self::check_keyword(&ident) {
                    Some(kw) => kw.to_string(),
                    None => "ID".to_string(),
                };

                tokens.push(Token {
                    r#type: token_type,
                    lexeme: ident,
                    line,
                    col: start_col,
                });

                continue;
            }

            // 3. Reconocimiento de Cadenas de Texto / Strings ("..." o '...')
            if c == '"' || c == '\'' {
                let quote = c;
                let start_col = col;
                let start_line = line;
                let mut str_val = String::new();
                str_val.push(quote);
                i += 1;
                col += 1;

                let mut closed = false;

                while i < len {
                    let current = chars[i];

                    // Manejo de caracteres de escape (\", \n, \t, \\, etc.)
                    if current == '\\' {
                        str_val.push('\\');
                        i += 1;
                        col += 1;
                        if i < len {
                            str_val.push(chars[i]);
                            i += 1;
                            col += 1;
                        }
                        continue;
                    }

                    // Salto de línea inesperado dentro de la cadena
                    if current == '\n' {
                        break;
                    }

                    // Cierre de la cadena
                    if current == quote {
                        str_val.push(quote);
                        i += 1;
                        col += 1;
                        closed = true;
                        break;
                    }

                    str_val.push(current);
                    i += 1;
                    col += 1;
                }

                if closed {
                    tokens.push(Token {
                        r#type: "STRING".to_string(),
                        lexeme: str_val,
                        line: start_line,
                        col: start_col,
                    });
                } else {
                    errors.push(LexerError {
                        line: start_line,
                        col: start_col,
                        lexeme: str_val,
                        message: format!("Cadena de texto no cerrada con comilla ({})", quote),
                    });
                }

                continue;
            }

            // 4. Reconocimiento de NUM_INT y NUM_DEC
            if c.is_ascii_digit() {
                let start_col = col;
                let mut num_str = String::new();
                let mut is_decimal = false;

                // Leer parte entera
                while i < len && chars[i].is_ascii_digit() {
                    num_str.push(chars[i]);
                    i += 1;
                    col += 1;
                }

                // Verificar si hay punto decimal seguido de AL MENOS un dígito
                if i < len && chars[i] == '.' {
                    if i + 1 < len && chars[i + 1].is_ascii_digit() {
                        is_decimal = true;
                        num_str.push('.'); // Agregar el punto
                        i += 1;
                        col += 1;

                        // Leer parte decimal
                        while i < len && chars[i].is_ascii_digit() {
                            num_str.push(chars[i]);
                            i += 1;
                            col += 1;
                        }
                    }
                }

                let token_type = if is_decimal { "NUM_DEC" } else { "NUM_INT" };

                tokens.push(Token {
                    r#type: token_type.to_string(),
                    lexeme: num_str,
                    line,
                    col: start_col,
                });

                continue;
            }

            // 5. Todo carácter no reconocido se reporta como Error Léxico
            errors.push(LexerError {
                line,
                col,
                lexeme: c.to_string(),
                message: format!("Símbolo no reconocido: '{}'", c),
            });

            i += 1;
            col += 1;
        }

        LexerOutput { tokens, errors }
    }
}

File: src/Lexer.cpp
#include "lexer.h"
#include <unordered_map>
#include <sstream>

// Equivalente a lexer.rs — Lexer::check_keyword y Lexer::analyze

/// Determina si un lexema es una palabra reservada y devuelve su tipo de token.
std::optional<std::string> Lexer::check_keyword(const std::string& lexeme) {
    static const std::unordered_map<std::string, std::string> keywords = {
        // Estructuras de control
        {"if",       "KW_IF"},
        {"else",     "KW_ELSE"},
        {"while",    "KW_WHILE"},
        {"for",      "KW_FOR"},
        {"do",       "KW_DO"},
        {"return",   "KW_RETURN"},
        {"break",    "KW_BREAK"},
        {"continue", "KW_CONTINUE"},
        {"switch",   "KW_SWITCH"},
        {"case",     "KW_CASE"},
        {"default",  "KW_DEFAULT"},

        // Declaraciones y tipos de datos
        {"let",      "KW_LET"},
        {"var",      "KW_VAR"},
        {"const",    "KW_CONST"},
        {"fn",       "KW_FN"},
        {"function", "KW_FUNCTION"},
        {"int",      "KW_INT"},
        {"float",    "KW_FLOAT"},
        {"double",   "KW_DOUBLE"},
        {"char",     "KW_CHAR"},
        {"string",   "KW_STRING"},
        {"bool",     "KW_BOOL"},
        {"void",     "KW_VOID"},
        {"struct",   "KW_STRUCT"},
        {"class",    "KW_CLASS"},

        // Literales booleanos / nulos
        {"true",     "KW_TRUE"},
        {"false",    "KW_FALSE"},
        {"null",     "KW_NULL"},

        // Funciones / identificadores clave comunes
        {"print",    "KW_PRINT"},
        {"println",  "KW_PRINTLN"},
        {"main",     "KW_MAIN"},
    };

    auto it = keywords.find(lexeme);
    if (it != keywords.end()) {
        return it->second;
    }
    return std::nullopt;
}

/// Ejecuta el análisis léxico sobre el texto de entrada.
LexerOutput Lexer::analyze(const std::string& input) {
    std::vector<Token> tokens;
    std::vector<LexerError> errors;

    const auto& chars = input;
    int len = static_cast<int>(chars.size());

    int i = 0;
    int line = 1;
    int col = 1;

    while (i < len) {
        char c = chars[i];

        // 1. Manejo de saltos de línea y espacios en blanco
        if (c == '\n') {
            line++;
            col = 1;
            i++;
            continue;
        } else if (c == '\r') {
            // Ignorar retorno de carro (para compatibilidad Windows \r\n)
            i++;
            continue;
        } else if (c == ' ' || c == '\t') {
            col++;
            i++;
            continue;
        }

        // 2. Reconocimiento de Identificadores (ID) y Palabras Reservadas (KW_*)
        if (std::isalpha(static_cast<unsigned char>(c)) || c == '_') {
            int start_col = col;
            std::string ident;

            while (i < len && (std::isalnum(static_cast<unsigned char>(chars[i])) || chars[i] == '_')) {
                ident += chars[i];
                i++;
                col++;
            }

            // Discriminación: Palabra Reservada vs Identificador
            auto kw = check_keyword(ident);
            std::string token_type = kw.has_value() ? kw.value() : "ID";

            tokens.push_back(Token{token_type, ident, line, start_col});
            continue;
        }

        // 3. Reconocimiento de Cadenas de Texto / Strings ("..." o '...')
        if (c == '"' || c == '\'') {
            char quote = c;
            int start_col = col;
            int start_line = line;
            std::string str_val;
            str_val += quote;
            i++;
            col++;

            bool closed = false;

            while (i < len) {
                char current = chars[i];

                // Manejo de caracteres de escape (\", \n, \t, \\, etc.)
                if (current == '\\') {
                    str_val += '\\';
                    i++;
                    col++;
                    if (i < len) {
                        str_val += chars[i];
                        i++;
                        col++;
                    }
                    continue;
                }

                // Salto de línea inesperado dentro de la cadena
                if (current == '\n') {
                    break;
                }

                // Cierre de la cadena
                if (current == quote) {
                    str_val += quote;
                    i++;
                    col++;
                    closed = true;
                    break;
                }

                str_val += current;
                i++;
                col++;
            }

            if (closed) {
                tokens.push_back(Token{"STRING", str_val, start_line, start_col});
            } else {
                std::string msg = "Cadena de texto no cerrada con comilla (";
                msg += quote;
                msg += ")";
                errors.push_back(LexerError{start_line, start_col, str_val, msg});
            }

            continue;
        }

        // 4. Reconocimiento de NUM_INT y NUM_DEC
        if (std::isdigit(static_cast<unsigned char>(c))) {
            int start_col = col;
            std::string num_str;
            bool is_decimal = false;

            // Leer parte entera
            while (i < len && std::isdigit(static_cast<unsigned char>(chars[i]))) {
                num_str += chars[i];
                i++;
                col++;
            }

            // Verificar si hay punto decimal seguido de AL MENOS un dígito
            if (i < len && chars[i] == '.') {
                if (i + 1 < len && std::isdigit(static_cast<unsigned char>(chars[i + 1]))) {
                    is_decimal = true;
                    num_str += '.'; // Agregar el punto
                    i++;
                    col++;

                    // Leer parte decimal
                    while (i < len && std::isdigit(static_cast<unsigned char>(chars[i]))) {
                        num_str += chars[i];
                        i++;
                        col++;
                    }
                }
            }

            std::string token_type = is_decimal ? "NUM_DEC" : "NUM_INT";
            tokens.push_back(Token{token_type, num_str, line, start_col});
            continue;
        }

        // 5. Todo carácter no reconocido se reporta como Error Léxico
        std::string msg = "Símbolo no reconocido: '";
        msg += c;
        msg += "'";
        errors.push_back(LexerError{line, col, std::string(1, c), msg});

        i++;
        col++;
    }

    return LexerOutput{tokens, errors};
}
